### PR TITLE
added Page 1 of Chapter 11 Tic tac toe project (#3 #92)

### PR DESCRIPTION
**Pull request by spamegg1**
see #3 and #92

### CODE REVIEW

**Comment by alexmojaki** [comment]
Please also sort out https://github.com/alexmojaki/futurecoder/pull/106#issuecomment-735565744

**Comment by spamegg1** [comment]
Oh right! It was on my mind but I forgot

File: backend/main/chapters/c11_tic_tac_toe_project.py [hunk 1]
```diff
@@ -0,0 +1,606 @@
...
+
+        hints = """
+You need to check every row, so you'll need a loop.
+How can you check if all entries in a row are equal to each other?
+If you don't know the size of a row, you'll have to loop all the way through it.
+So you'll end up using nested loops!
+For each row, define a boolean. Then loop through that row, updating the boolean accordingly.
+You can use the first entry `row[0]` in a row to compare all the row entries to it.
+Think carefully about what the initial value of the boolean should be, and under what conditions you should change its value.
+After looping through a row, if you determined that all its entries are equal, then return `True` (ending the outer loop early).
```
Comment (alexmojaki) [inline backend/main/chapters/c11_tic_tac_toe_project.py:127]: Need to cover this concept in the functions chapter

Comment (spamegg1) [reply backend/main/chapters/c11_tic_tac_toe_project.py:127]: OK. Will get to it. Would `Returning Values from Functions` be the best place for it? Would one `VerbatimStep` suffice? Or should I make a new page with many steps?

Comment (spamegg1) [reply backend/main/chapters/c11_tic_tac_toe_project.py:127]: On second thought, I should add at least 2 or 3 steps. I should show one `if-else` block example, maybe mention that `else` is unnecessary; and I should show one `for` loop example, with Snoop, so they can see the loop doesn't go all the way.

**Comment by alexmojaki** [comment]
I think the early return concept needs a new page at the end of the chapter, a few steps, and a markdown draft before coding. Here are some ideas:

Explain what happens here:

```python
def foo():
    return 1
    return 2

print(foo())
```

Sometimes beginners try returning many values with something like this:

```python
def double_numbers(numbers):
    for x in numbers:
        return x * 2

assert_equal(double_numbers([1, 2, 3]), [2, 4, 6])
```

Explain why this doesn't work and that they need to create a list and append to it instead.

Emphasise that `return` ends the whole function, not just the loop. Contrast this:

```python
def foo():
    for letter in 'abcde':
        for number in range(3):
            print(f"{letter} {number}")
            if letter == 'c':
                return letter
foo()
```

with using `break` instead of `return letter`, which only breaks the inner loop.

**Comment by alexmojaki** [comment]
There's still a lot of duplicated `generate_inputs` code. A higher order function will work well to reduce that.

**Comment by spamegg1** [comment]
Something like this? I didn't know where else to put `generate_board`, should it go into `main/exercises.py`?

**Comment by spamegg1** [comment]
The changes look good by the way, text and hints are clearer :+1:

File: backend/main/chapters/c11_tic_tac_toe_project.py [hunk 2]
```diff
@@ -7,6 +7,31 @@
 from main.text import ExerciseStep, Page, MessageStep, Disallowed
 
 
+def generate_board(board_type):
+    winning, size = choice([True, False]), randint(3, 10)
+    char1, char2 = choice(ascii_uppercase), choice(ascii_uppercase)
```
Comment (alexmojaki) [inline backend/main/chapters/c11_tic_tac_toe_project.py:12]: I don't like this style, please just assign one per line like before.

Comment (spamegg1) [reply backend/main/chapters/c11_tic_tac_toe_project.py:12]: OK.

**Comment by alexmojaki** [comment]
`generate_board` is perfect, even better than what I had in mind. No reason to move it anywhere.

**Comment by spamegg1** [comment]
Well then I'll get to work on that draft

**Pull request merged**